smartsim/entity/strategies.py

```diff
@@ -90,32 +95,69 @@ def _impl(
 # single application if parameters only have one value
 @_register("all_perm")
 def create_all_permutations(
-    params: t.Mapping[str, t.Sequence[str]],
+    file_params: t.Mapping[str, t.Sequence[str]],
+    exe_arg_params: t.Mapping[str, t.Sequence[t.Sequence[str]]],
     _n_permutations: int = 0,
-    # ^^^^^^^^^^^^^
-    # TODO: Really don't like that this attr is ignored, but going to leave it
-    #       as the original impl for now. Will change if requested!
-) -> list[dict[str, str]]:
-    permutations = itertools.product(*params.values())
-    return [dict(zip(params, permutation)) for permutation in permutations]
+) -> list[ParamSet]:
+    # Generate all possible permutations of parameter values
+    file_params_permutations = itertools.product(*file_params.values())
+    # Create dictionaries for each parameter permutation
+    param_zip = [
+        dict(zip(file_params, permutation)) for permutation in file_params_permutations
+    ][:_n_permutations]
+    # Generate all possible permutations of executable arguments
+    exe_arg_params_permutations = itertools.product(*exe_arg_params.values())
+    # Create dictionaries for each executable argument permutation
+    exe_arg_zip = [
+        dict(zip(exe_arg_params, permutation))
+        for permutation in exe_arg_params_permutations
+    ][:_n_permutations]
+    # Combine parameter and executable argument dictionaries
+    combinations = itertools.product(param_zip, exe_arg_zip)
+    # Combine the parameter sets from 'param_zip' and 'exe_arg_zip' using itertools.zip_longest
+    param_set = (ParamSet(file_param, exe_arg) for file_param, exe_arg in combinations)
+    slice = itertools.islice(param_set, _n_permutations)
+    return list(slice)
 
 
 @_register("step")
 def step_values(
-    params: t.Mapping[str, t.Sequence[str]], _n_permutations: int = 0
-) -> list[dict[str, str]]:
-    steps = zip(*params.values())
-    return [dict(zip(params, step)) for step in steps]
+    params: t.Mapping[str, t.Sequence[str]],
+    exe_args: t.Mapping[str, t.Sequence[t.Sequence[str]]],
+    _n_permutations: int = 0,
```

Comment:
same naming nit here:
 ```suggestion
    n_permutations: int = 0,
```